datafusion/common/src/config.rs

```diff
@@ -768,7 +768,18 @@ impl ConfigField for ConfigOptions {
         let (key, rem) = key.split_once('.').unwrap_or((key, ""));
         match key {
             "catalog" => self.catalog.set(rem, value),
-            "execution" => self.execution.set(rem, value),
+            "execution" => {
+                let value = match (rem, value) {
+                    ("target_partitions", "0") => {
+                        &ExecutionOptions::default().target_partitions.to_string()
+                    }
+                    ("planning_concurrency", "0") => {
+                        &ExecutionOptions::default().planning_concurrency.to_string()
+                    }
+                    _ => value,
+                };
+                self.execution.set(rem, value)
+            }
```

Comment:
I'd make this a transform - a single dedicated function should work, since both `target_partitions` and `planning_concurrency` default to `get_available_parallelism`

Something like:
```rs
fn normalized_parallelism(n: &str) -> String {
    if n.parse::<usize>() == Ok(0) {
        get_available_parallelism().to_string()
    } else {
        n.to_owned()
    }
}
```